### PR TITLE
LibGfx: Make JPGLoader iterate components deterministically

### PR DESCRIPTION
**Pull request by sin-ack**
`JPGLoader` used to store component information in a `HashTable`, indexed
by the ID assigned by the JPEG file.  This was fine for most purposes,
however after f89e8fb7 this was revealed to be a flawed implementation
which causes non-deterministic iteration over components.

This issue was previously masked by a perfect storm of `int_hash` being
stable for the integer values 0, 1 and 2; and `AK::HashTable` having just
the right amount of buckets for the components to be ordered correctly
after being hashed with int_hash. However, after f89e8fb7,
`malloc_good_size` was used for determining the amount of space for
allocation; this caused the ordering of the components to change, and
images started showing up with the red and blue channels reversed. The
issue was finally determined to be inconsistent ordering after randomly
changing the order of the components caused Huffman decoding to fail.

This was the result of about 10 hours of hair-pulling and repeatedly
doing full rebuilds due to bisecting between commits that touched AK.
Gunnar, I like you, but please don't make me go through this again. :^)

Credits to Andrew Kaster, bgianf, CxByte and Gunnar for the debugging
help.

### CODE REVIEW

File: Userland/Libraries/LibGfx/JPGLoader.cpp [hunk 1]
```diff
@@ -502,21 +513,14 @@ static bool read_start_of_scan(InputMemoryStream& stream, JPGLoadingContext& con
...
-            component = &it->value;
-            if (i != component->serial_id) {
-                dbgln("JPEG decode failed (i != component->serial_id)");
-                return false;
-            }
-        } else {
-            dbgln_if(JPG_DEBUG, "{}: Unsupported component id: {}!", stream.offset(), component_id);
+        auto& component = context.components[i];
+        if (component.id != component_id) {
+            dbgln("JPEG decode failed (i != component->serial_id)");
```
Comment (linusg) [inline Userland/Libraries/LibGfx/JPGLoader.cpp:523]:
```suggestion
            dbgln("JPEG decode failed (component.id != component_id)");
```

**Review by linusg — CHANGES_REQUESTED** [review]

**Comment by linusg** [comment]
Nicely done everyone involved!

**Comment by wezm** [comment]
I note that in https://github.com/SerenityOS/serenity/commit/f107c7065230c6ffa179caca28773e04d4f3107d (part of PR https://github.com/SerenityOS/serenity/pull/4237) `components` was deliberately changed from a `Vector` to a `HashMap` to handle an unlikely edge case, is that still an issue?

**Comment by riking** [comment]
You should change the standard library to randomize hash table iteration order, to prevent any recurrence of this type of issue in the future.

**Pull request merged**